### DISCUSSION TITLE
Custom Login  Parameter and custom admin views

### DISCUSSION THREAD

**Discussion by Fidel-C** [piccolo-orm/piccolo_api#225] [General]
I'd like to know if there's an easy way to select a particular login field, such as email or phone number instead of username which is the default login field or even totally exclude some fields from the BaseUser model without having to reinvent the wheel?

Secondly, I'd like to know whether there's a way to add custom views to piccolo admin to display custom widgets like graphs and charts if need be.

**Comment by sinisaos** [comment]
> I'd like to know if there's an easy way to select a particular login field, such as email or phone number instead of username which is the default login field or even totally exclude some fields from the BaseUser model without having to reinvent the wheel?
> 

There is already an issue for that [here](https://github.com/piccolo-orm/piccolo_api/issues/165) and you can try some of the solutions offered. The main problem is that `BaseUser` has two required fields **username** and **password** ([create_user](https://github.com/piccolo-orm/piccolo/blob/master/piccolo/apps/user/tables.py#L244) method) and `session auth` depends on `BaseUser`. Not the best solution, but maybe you can try to make a custom login template and use phone or email as username (username is a string type and you can pass any string to it).

> Secondly, I'd like to know whether there's a way to add custom views to piccolo admin to display custom widgets like graphs and charts if need be.

[Here](https://github.com/piccolo-orm/piccolo_admin/issues/146) is the issue for that, feel free to implement the solution.

**Reply by Fidel-C** [reply]
@sinisaos thanks for your reply.

**Comment by dantownsend** [comment]
> I'd like to know if there's an easy way to select a particular login field, such as email or phone number instead of username which is the default login field

We could let the user configure which column to use - for example `email` instead of `username`, but all of the login endpoints in Piccolo API use `BaseUser.login` under the hood which currently expects a username. We would have to refactor that too, so `email` could be used instead.

> exclude some fields from the BaseUser model without having to reinvent the wheel?

Excluding fields from `BaseUser` is a bit tricky. Adding them is easy using class inheritance.

Most of the Piccolo API endpoints and middleware allow you to pass in an alternative to `BaseUser`, for example the `auth_table` argument of `session_login`:

https://piccolo-api.readthedocs.io/en/latest/session_auth/endpoints.html#source

There's nothing stopping you from passing in your own alternative implementation to `BaseUser`. It just needs to implement an async `login` method, and have a `username` and `password` column as a minimum.

> Secondly, I'd like to know whether there's a way to add custom views to piccolo admin to display custom widgets like graphs and charts if need be.

Like @sinisaos said, this is something we'd love to add, but not currently possible.

**Reply by Fidel-C** [reply]
@dantownsend thanks a lot
